File: 1/2021.12.01/1584.Min Cost to Connect All Points/python/solution.py
class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        n = len(points)
        if n < 2:
            return 0
        arr = []
        for i in range(n - 1):
            p1 = points[i]
            for j in range(1, n):
                p2 = points[j]
                arr.append((i, j, abs(p1[0] - p2[0]) + abs(p1[1] - p2[1])))
        arr.sort(key=lambda x: x[2])
        mst = 0
        uf = UF(n)
        for i, j, weight in arr:
            if uf.is_connected(i, j):
                continue
            mst += weight
            uf.union(i, j)
            if uf.count == 1:
                return mst



class UF:

    def __init__(self, count) -> None:
        self.count = count
        self.parent = [i for i in range(count)]
        self.size = [1] * count
    

    def union(self, p, q):
        root_p, root_q = self._find(p), self._find(q)
        if root_p == root_q:
            return
        if self.size[root_p] > self.size[root_q]:
            self.parent[root_q] = root_p
            self.size[root_p] += self.size[root_q]
        else:
            self.parent[root_p] = root_q
            self.size[root_q] += self.size[root_p]
        self.count -= 1


    def is_connected(self, p, q):
        root_p, root_q = self._find(p), self._find(q)
        return root_p == root_q

    def _find(self, x):
        while self.parent[x] != x:
            self.parent[x] = self.parent[self.parent[x]]
            x = self.parent[x]
        return x

File: 1/2021.12.01/1584.Min Cost to Connect All Points/java/solution.java
class Solution {
    public int minCostConnectPoints(int[][] points) {
        int n = points.length;
        if (n < 2) {
            return 0;
        }
        int[][] arr = new int[n * (n - 1) / 2][];
        int idx = 0;
        for (int i = 0; i < n - 1; ++i) {
            int[] p1 = points[i];
            for (int j = i + 1; j < n; ++j) {
                int[] p2 = points[j];
                arr[idx++] = new int[]{i, j, Math.abs(p1[0] - p2[0]) + Math.abs(p1[1] - p2[1])};
            }
        }
        Arrays.sort(arr, (a, b) -> a[2] - b[2]);
        int mst = 0;
        UF uf = new UF(n);
        for (int[] item : arr) {
            int p1 = item[0];
            int p2 = item[1];
            int edge = item[2];
            if (uf.isConnected(p1, p2)) {
                continue;
            }
            uf.union(p1, p2);
            mst += edge;
            if (uf.count() == 1) {
                return mst;
            }
        }
        return mst;
    }
}

class UF {

    private int[] parent;
    private int count;
    private int[] size;

    public UF(int count) {
        this.count = count;
        parent = new int[count];
        size = new int[count];
        for (int i = 0; i < count; ++i) {
            parent[i] = i;
            size[i] = 1;
        }
    }

    public void union(int p, int q) {
        int rootP = find(p);
        int rootQ = find(q);
        if (rootP == rootQ) {
            return;
        }
        if (size[rootP] > size[rootQ]) {
            parent[rootQ] = rootP;
            size[rootP] += size[rootQ];
        } else {
            parent[rootP] = rootQ;
            size[rootQ] += size[rootP];
        }
        --count;
    }

    public boolean isConnected(int p, int q) {
        int rootP = find(p);
        int rootQ = find(q);
        return rootP == rootQ;
    }

    private int find(int x) {
        while (parent[x] != x) {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    }

    public int count() {
        return count;
    }

}
